File: linked_list/2095_delete_the_middle_of_a_linked_list.py
from typing import Optional

class ListNode:
    def __init__(self, val=0):
        self.val = val
        self.next = next


def delete_middle(head: Optional[ListNode]) -> Optional[ListNode]:
    dummy = ListNode()
    dummy.next = head

    slow = dummy
    fast = dummy.next

    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next

    slow.next = slow.next.next
    return dummy.next

File: linked_list/2095_delete_the_middle_of_a_linked_list.go
type ListNode struct {
	Val  int
	Next *ListNode
}

// по условию head всегда есть
func deleteMiddle(head *ListNode) *ListNode {
	// проверяем что есть 2 ноды
	if head.Next == nil {
		// деволтное значение указателя nil
		return nil
	}
	slow := head
	// даем фору
	fast := head.Next.Next

	for fast != nil && fast.Next != nil {
		fast = fast.Next.Next
		slow = slow.Next
	}

	slow.Next = slow.Next.Next
	return head
}
